docs/articles/documentation/recipes/best-practices/best-practices.md

```diff
@@ -0,0 +1,223 @@
+
+> The smart assertion query mechanism works with promises generated by TestCafe Selector API. Use the `await` keyword with user-created promises and promises generated by third-party libraries.
+
+This rule applies to ClientFunctions in the same capacity.
+
+The test below doesn't yield consistent results:
+
+```js
+import { Selector, ClientFunction } from 'testcafe';
+
+fixture `My fixture`
+    .page `http://devexpress.github.io/testcafe/example/`;
+
+test('Assertion with ClientFunction', async t => {
+    const getValue        = ClientFunction(() => document.getElementById('preferred-interface').value)
+    const interfaceSelect = Selector('#preferred-interface');
+    const interfaceOption = interfaceSelect.find('option');
+    const value           = await getValue();
+
+
+    await t
+        .click(interfaceSelect)
+        .click(interfaceOption.withText('JavaScript API'))
+        .expect(value).eql('JavaScript API')
+});
+```
+
+In this example, the client function obtains the value of a drop-down element. This value is then passed to the `t.expect` method. Since the client function is awaited when it's assigned to the variable, the value gets resolved instantly and never updates. The smart assertion query mechanism doesn't apply and the test fails.
+
+To solve the issue, pass the client function without the `await` keyword:
+
+```js
+import { Selector, ClientFunction } from 'testcafe';
+
+fixture `My fixture`
+    .page `http://devexpress.github.io/testcafe/example/`;
+
+test('Assertion with ClientFunction', async t => {
+    const getValue = ClientFunction(() => document.getElementById('preferred-interface').value)
+
+    const interfaceSelect = Selector('#preferred-interface');
+    const interfaceOption = interfaceSelect.find('option');
+    const value           =  getValue();
+
+    await t
+        .click(interfaceSelect)
+        .click(interfaceOption.withText('JavaScript API'))
+        .expect(value).eql('JavaScript API')
+});
+```
+
+In this case, TestCafe applies the smart query mechanism and retries the assertion multiple times. This test responds to the changes that happen on the page and is more conclusive.
+
+## Use of Page Objects
+
+Use the Page Model in your tests. It helps you abstract the tested page into a single file that you can use to target page elements.
+
+This approach improves the flexibility of your tests - if the UI changes, change one file to update all the references. Define common operations as your model's methods. Tests are then more readable and less brittle because there is no duplicate code.
+
+The page model use is described in more detail in the [Page Model](../../guides/concepts/page-model.md) article.
+
+A code example is available in the [testcafe-examples](https://github.com/DevExpress/testcafe-examples/tree/master/examples/use-page-model) repository.
+
+## Use of Roles for Login
```

Comment:
Same thing about code samples here. This section will be inevitably lost, it's just one and a half sentences.

We can say at least a couple of sentences and show some code, and only then link to the dedicated guide for details.